mlflow/tracing/clients/ipython_wrapper.py

```diff
@@ -0,0 +1,42 @@
+from trace import Trace
+
+from mlflow.tracing.clients.base import TraceClient
+from mlflow.tracing.clients.local import InMemoryTraceClient
+
+
+class IPythonTraceWrapper(InMemoryTraceClient):
+    """
+    Trace client that wraps the InMemoryTraceClient
+    in IPython notebooks. The only purpose of this
+    class is to add a hook on log_trace() to display
+    the contents of the trace.
+    """
+
+    _instance: TraceClient = None
+
+    @classmethod
+    def get_instance(cls) -> InMemoryTraceClient:
+        if cls._instance is None:
+            cls._instance = IPythonTraceWrapper()
+        return cls._instance
+
+    def __init__(self):
+        self.prev_execution_count = -1
+        super().__init__()
+
+    def log_trace(self, trace: Trace):
+        from IPython import get_ipython
+        from IPython.display import display
+
+        # check the current exec count to know if the user is
+        # running the command in a new cell or not. this is
+        # useful because the user might generate multiple traces
+        # in a single cell, and we only want to display the last one.
+        current_exec_count = get_ipython().execution_count
+        if self.prev_execution_count != current_exec_count:
+            self.prev_execution_count = current_exec_count
+            self.display_handle = display(trace, display_id=True)
+        else:
+            self.display_handle.update(trace)
```

Comment:
So the displayed item might be updated multiple times. Is it possible to only trigger on `display` on the last trace item without invoking `update` ?

I remember the `cell` has an execution completion hook, maybe we can move display to there ?

Reply:
i'll investigate!

Reply:
i wonder if it's a bit messy to put it in there though—it seems like we need to register a callback, like `print_last_trace()` that will execute on every cell completion. however there might be many cells in which the user is not executing trace-related code, and in order to detect those we might need to check the contents of the cell.

is there any specific concern with having this logic in the `log_trace()` function?

Reply:
OK. Let's keep current PR's code if it doesn't cause performance issue when many updates occurs in one cell